test/distributed/fsdp/test_fsdp_comm_hooks.py

```diff
@@ -0,0 +1,239 @@
+        Arguments:
+            has_wrapping (bool): Configures wrapping of a module.
+            sharding_strategy (Optional[ShardingStrategy]): Configures the FSDP algorithm.
+        """
+
+        class DummyState(object):
+
+            __slots__ = [
+                "process_group"
+            ]
+
+            def __init__(self, process_group):
+                self.process_group = process_group
+
+        def dummy_hook(state: DummyState, grad: torch.Tensor):
+            pass
+
+
+        class Net(nn.Module):
+
+            def __init__(self, has_wrapping):
+                # to ensure determinizm
+                torch.manual_seed(0)
+                torch.cuda.manual_seed(0)
+                super().__init__()
+
+                if has_wrapping:
+                    self.net = FSDP(nn.Sequential(
+                        nn.Linear(8, 16),
+                        nn.ReLU(),
+                        FSDP(
+                            nn.Linear(16, 8),
+                            device_id=torch.cuda.current_device(),
+                            sharding_strategy=sharding_strategy
+                        )
+                    ),
+                        device_id=torch.cuda.current_device(),
+                        sharding_strategy=sharding_strategy
+                    )
+                else:
+                    self.net = nn.Sequential(
+                        nn.Linear(8, 16),
+                        nn.ReLU(),
+                        nn.Linear(16, 8)
+                    )
+
+                self.out = nn.Linear(8, 4)
+
+            def forward(self, x):
+                return self.out(F.relu(self.net(x)))
+
+        # Initialize the model and inputs
+        device = torch.device("cuda")
+        fsdp_model_with_hook = FSDP(
+            Net(has_wrapping),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy
+        ).to(device)
+
+        dummy_state = DummyState(process_group=None)
+
+        # FSDP currently suports communication hooks for a NO_SHARD strategy
+        # Check that a `NotImplementedError`` is raised for other strategies
+        if sharding_strategy != ShardingStrategy.NO_SHARD:
+            # Check that default hook is set to None
+            for entry in FSDP.fsdp_modules(fsdp_model_with_hook):
+                self.assertIsNone(entry.communication_hook)
+                self.assertIsNone(entry.communication_hook_state)
+
+            with self.assertRaises(
+                NotImplementedError,
+                msg="Communication hooks are currently only available for a NO_SHARD strategy."
+            ):
+                fsdp_model_with_hook.register_comm_hook(dummy_state, dummy_hook)
+
+        else:
+
+            # Check that default hook is set to `all_reduce`
+            for entry in FSDP.fsdp_modules(fsdp_model_with_hook):
+                self.assertEqual(entry.communication_hook.__qualname__, allreduce_hook.allreduce_hook.__qualname__)
```

Comment:
Curious: Is it possible to check the function reference itself?
```suggestion
                self.assertEqual(entry.communication_hook, allreduce_hook.allreduce_hook)
```